base/sparse/cholmod.jl

```diff
@@ -231,836 +181,892 @@ else
-size(B::CholmodDense,d) = size(B.mat,d)
-size(A::CholmodSparse) = (int(A.c.m), int(A.c.n))
-function size(A::CholmodSparse, d::Integer)
-    d == 1 ? A.c.m : (d == 2 ? A.c.n : 1)
+function convert{Tv,Ti}(::Type{SparseMatrixCSC{Tv,Ti}}, A::Sparse{Tv,Ti})
+    s = unsafe_load(A.p)
+    B = SparseMatrixCSC(s.nrow, s.ncol, increment(pointer_to_array(s.p, (s.ncol + 1,), false)), increment(pointer_to_array(s.i, (s.nzmax,), false)), copy(pointer_to_array(s.x, (s.nzmax,), false)))
+    if s.stype != 0
+        return isreal(A) ? Symmetric(B, s.stype < 0 ? :L : :U) : Hermitian(B, s.stype < 0 ? :L : :U)
+    end
+    return B
+end
+sparse{Tv,Ti}(A::Sparse{Tv,Ti}) = convert(SparseMatrixCSC{Tv,Ti}, A)
+sparse(L::Factor)   = sparse(Sparse(L))
+sparse(D::Dense)    = sparse(Sparse(D))
+function full(A::Sparse)
+    B = convert(Matrix, Dense(A))
+    s = unsafe_load(A.p)
+    if s.stype != 0
+        return isreal(A) ? Symmetric(B, s.stype < 0 ? :L : :U) : Hermitian(B, s.stype < 0 ? :L : :U)
+    end
+    return B
+end
+full(A::Dense) = convert(Matrix, A)
+
+# Calculate the offset into the stype field of the cholmod_sparse_struct and
+# change the value
+function change_stype!(A::Sparse, i::Integer)
+    offset = fieldoffsets(C_Sparse)[names(C_Sparse) .== :stype][1]
+    unsafe_store!(convert(Ptr{Cint}, A.p), i, div(offset, 4) + 1)
+    nothing
 end
-size(L::CholmodFactor) = (n = int(L.c.n); (n,n))
-size(L::CholmodFactor,d::Integer) = d < 1 ? throw(BoundsError("Dimension $d out of range")) : (d <= 2 ? int(L.c.n) : 1)
 
-function solve{Tv<:CHMVTypes,Ti<:CHMITypes}(L::CholmodFactor{Tv,Ti},
-                                            B::SparseMatrixCSC{Tv,Ti},typ::Integer)
-    sparse!(solve(L,CholmodSparse(B),typ))
+free!(A::Dense) = free_dense(A.p)
+free!(A::Sparse) = free_sparse(A.p)
+free!(F::Factor) = free_factor(F.p)
+
+eltype{T<:VTypes}(A::Dense{T}) = T
+eltype{T<:VTypes}(A::Factor{T}) = T
+eltype{T<:VTypes}(A::Sparse{T}) = T
+
+function show(io::IO, F::Factor)
+    s = unsafe_load(F.p)
+    println(io, typeof(F))
+    @printf(io, "type: %12s\n", bool(s.is_ll) ? "LLt" : "LDLt")
+    @printf(io, "method: %10s\n", bool(s.is_super) ? "supernodal" : "simplicial")
+    @printf(io, "maxnnz: %10d\n", int(s.nzmax))
+    @printf(io, "nnz: %13d\n", nnz(Sparse(F)))
 end
-function solve{Tv<:CHMVTypes,Ti<:CHMITypes}(L::CholmodFactor{Tv,Ti},B::SparseMatrixCSC{Tv,Ti})
-    sparse!(solve(L,CholmodSparse(B),CHOLMOD_A))
+
+isvalid(A::Dense) = check_dense(A)
+isvalid(A::Sparse) = check_sparse(A)
+isvalid(A::Factor) = check_factor(A)
+
+copy(A::Dense) = copy_dense(A)
+copy(A::Sparse) = copy_sparse(A)
+copy(A::Factor) = copy_factor(A)
+
+function size(A::Union(Dense,Sparse))
+    s = unsafe_load(A.p)
+    return (int(s.nrow), int(s.ncol))
 end
-function solve{Tv<:CHMVTypes,Ti<:CHMITypes}(L::CholmodFactor{Tv,Ti},B::CholmodSparse{Tv,Ti})
-    solve(L,B,CHOLMOD_A)
+function size(F::Factor, i::Integer)
+    i < 1 && error("arraysize: dimension out of range")
```

Comment:
I've tried to mimic the behavior of
```julia
julia> size(randn(3,3), 0)
ERROR: arraysize: dimension out of range
 in size at array.jl:19
```
but I don't know if that is the right thing to do.